File: advent_of_code_2021/day10/part_2.py
#!/usr/bin/env python3

import sys
import math

from math import gcd, floor, sqrt, log
from collections import defaultdict, deque
from bisect import bisect_left, bisect_right

MOD = 1000000007

inputname = "real"
inputname = "example"

correct = {
    "(": ")",
    "[": "]",
    "{": "}",
    "<": ">",
}

_correct = {v: k for k, v in correct.items()}

errorscore = {
    ")": 1,
    "]": 2,
    "}": 3,
    ">": 4,
}

lines = open(inputname, "r").read().splitlines()

stack = []

score = 0

newlines = []

for idx, line in enumerate(lines):
    bad = False
    for char in line:
        if char in "([{<":
            stack.append(char)
        else:
            popped = stack.pop()
            if popped != _correct[char]:
                bad = True
                print("Error", char)
                score += errorscore[char]
    if not bad:
        newlines.append(line)

print(score)
        
for line in newlines:
    print(line)

File: advent_of_code_2021/day10/part_2.c
#include <stdio.h>

#define LEFT(c)((c) == '(' || (c) == '[' || (c) == '{' || (c == '<'))

int stack[BUFSIZ], sp = 0;

char inverse[128] = { [')'] = '(', [']'] = '[', ['}'] = '{', ['>'] = '<' };
int errscore[128] = { ['('] = 1, ['['] = 2, ['{'] = 3, ['<'] = 4 };

int main(void)
{
    int score = 0;
    char buf[BUFSIZ];
    FILE *input = fopen("real", "r");
    while (fgets(buf, BUFSIZ, input)) {
        for (char *s = buf; *s && *s != '\n'; s++) {
            if (LEFT(*s)) {
                stack[sp++] = *s;
            } else {
                if (stack[--sp] != inverse[*s]) {
                    score += errscore[*s];
                }
            }
        }
    }
    printf("%d\n", score);
}
/*
#!/usr/bin/env python3

inputname = "example"
inputname = "real"

inverse = { ")": "(", "]": "[", "}": "{", ">": "<" }
errorscore = { "(": 1, "[": 2, "{": 3, "<": 4 }

lines = open(inputname, "r").read().splitlines()

newlines = []

for idx, line in enumerate(lines):
    stack = []
    bad = False
    for char in line:
        if char in "([{<":
            stack.append(char)
        else:
            popped = stack.pop()
            if popped != inverse[char]:
                bad = True
    if not bad:
        newlines.append(line)

scores = []
for line in newlines:
    stack = []
    for char in line:
        if char in "([{<":
            stack.append(char)
        else:
            popped = stack.pop()
    score = 0
    while stack:
        score = score * 5 + errorscore[stack.pop()]
    scores.append(score)

print(sorted(scores)[len(scores) // 2])
*/
